### ISSUE TITLE
Patched method is not called

### ISSUE THREAD

**Issue by n3b** [pardeike/Harmony#363]
**Describe the bug**
I'm probably missing something, but can't make similar construction to work. Nested private class with public interface, seem to be patched normally, yet never called. 
And it works when interface is not used.

**To Reproduce**
```
using System;
using System.Reflection;
using HarmonyLib;

public class Test
{
    public static void Main()
    {
        var harmony = new Harmony("test");
        var source = AccessTools.Method(typeof(Parent).GetNestedType("Child", BindingFlags.NonPublic), "DoWork");
        var patch = AccessTools.Method(typeof(Test), nameof(Patch));
        harmony.Patch(source, new HarmonyMethod(patch));
        var parent = new Parent();
        parent.Start();
    }

    public static bool Patch()
    {
        var number = 5;
        
        int n1 = 0, n2 = 1, n3, i;
        Console.Write(n1 + " " + n2 + " ");
        for (i = 2; i < number; ++i)
        {
            n3 = n1 + n2;
            Console.Write(n3 + " ");
            n1 = n2;
            n2 = n3;
        }

        return false;
    }
}

public class Parent
{
    public void Start()
    {
        var child = new Child();
        chil.DoWork();
    }

    class Child : IChild
    {
        public void DoWork()
        {
            var number = 10;
            
            int n1 = 0, n2 = 1, n3, i;
            Console.Write(n1 + " " + n2 + " ");
            for (i = 2; i < number; ++i)
            {
                n3 = n1 + n2;
                Console.Write(n3 + " ");
                n1 = n2;
                n2 = n3;
            }
        }
    }
}

public interface IChild
{
    void DoWork();
}
```

**Expected behavior**
It has to output 5 fib numbers instead of 10.

**Runtime environment (please complete the following information):**
 - OS: Windows 10, 64bit
 - .NET Core 5.0
 - Harmony version 2.0.4

**Comment by pardeike** [comment]
Your patch is not executed because it is inlined. Try building it in debug mode to verify it.

**Comment by pardeike** [comment]
After some more testing, I can confirm that in .NET 5, methods that are implementing an interface are not patchable when optimizations are turned on:

```cs
using HarmonyLib;
using System;
using System.Collections.Generic;
using System.Reflection;

namespace Net5
{
	public class Net5Test
	{
		public static void Main()
		{
			var harmony = new Harmony("test");
			harmony.PatchAll();
			Parent.Start();
		}
	}

	[HarmonyPatch]
	internal static class Patch
	{
		internal static IEnumerable<MethodBase> TargetMethods()
		{
			var type = AccessTools.Inner(typeof(Parent), "Test");
			yield return AccessTools.Method(type, "Direct");
			yield return AccessTools.Method(type, "ViaInterface");
		}

		internal static void Postfix()
		{
			Console.WriteLine("PATCHED");
		}
	}

	public interface ITest
	{
		void ViaInterface();
	}

	public class Parent
	{
		public static void Start()
		{
			var child = new Test();

			child.ViaInterface();
			Console.WriteLine();

			child.Direct();
			Console.WriteLine();
		}

		class Test : ITest
		{
			// does patch in Release build
			public void Direct()
			{
				try { Console.Write("-> Direct: "); }
				catch (Exception e) { Console.Write(e); }
			}

			// cannot be patched in Release build
			public void ViaInterface()
			{
				try { Console.Write("-> ViaInterface: "); }
				catch (Exception e) { Console.Write(e); }
			}
		}
	}
}
```

**Comment by pardeike** [comment]
Keep in mind that Harmony officially does not support .NET 5 yet

**Comment by lbmaian** [comment]
Investigation in discord shows that this is the same problem as #337

**Comment by pardeike** [comment]
Closed in favor of #337
